osmoutils/store_helper.go

```diff
@@ -222,3 +224,61 @@ func DeleteAllKeysFromPrefix(store store.KVStore, prefixKey []byte) {
 		prefixStore.Delete(iter.Key())
 	}
 }
+
+// GetCoinArrayFromPrefix returns all coins from the store that has the given prefix.
+func GetCoinArrayFromPrefix(ctx sdk.Context, storeKey storetypes.StoreKey, storePrefix []byte) []sdk.Coin {
+	coinArray := make([]sdk.Coin, 0)
+
+	store := ctx.KVStore(storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, storePrefix)
+
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		bz := iterator.Value()
+		coin := sdk.Coin{}
+		if err := coin.Unmarshal(bz); err == nil {
+			coinArray = append(coinArray, coin)
+		}
+	}
+
+	return coinArray
+}
+
+// GetCoinByDenomFromPrefix returns the coin from the store that has the given prefix and denom.
+// If the denom is not found, a zero coin is returned.
+func GetCoinByDenomFromPrefix(ctx sdk.Context, storeKey storetypes.StoreKey, storePrefix []byte, denom string) (sdk.Coin, error) {
+	store := prefix.NewStore(ctx.KVStore(storeKey), storePrefix)
+	key := append(storePrefix, []byte(denom)...)
+
+	bz := store.Get(key)
+	if len(bz) == 0 {
+		return sdk.NewCoin(denom, osmomath.ZeroInt()), nil
+	}
+
+	coin := sdk.Coin{}
+	if err := coin.Unmarshal(bz); err != nil {
+		return sdk.NewCoin(denom, osmomath.ZeroInt()), err
+	}
+
+	return coin, nil
+}
```

Comment:
Shouldn't we only marshal the amount, since the key contains the denom?

Reply:
@ValarDragon Is there performance improvement if we just unmarshal the amount and then from that create a coin to return?

Reply:
yeah:
- less data written/read
- (and therefore gas)

Reply:
Two things:

1. Modified the key to use a separator for easier parsing of the denom from the key
2. Did not modify protorev's kvstore since we are not changing their store like we are the taker fee store. Otherwise I would need to recreate the store to store ints instead of coins, which feels out of scope of this PR

https://github.com/osmosis-labs/osmosis/pull/7240/commits/a27e3c852f156853590b50ba33ac4c501d328e18
https://github.com/osmosis-labs/osmosis/pull/7240/commits/a582fa14af24f2afd595cfc44eeb35bf5d837b0a